### PR TITLE
Undo more deletions, clear, add, edit

### LINKED ISSUE #62: Add Undo functionality for multiple deletions, similarly for Clear.
As a Financial Advisor, I would like the ability to undo my previous **deletions**, which can save me time for adding the person again. 
Currently, the undo feature can only undo the most recently deleted contact. If there are multiple deletions, the user cannot undo them all.

### LINKED ISSUE #80: Add Undo functionality for edit and add.
As a Financial Advisor, I would like the ability to undo my previous edits and additions, which can save me time for editing and deleting.

### PR DESCRIPTION
**Pull request by glenngnng**
Currently, the Undo command is only capable of undoing the most recently deleted contact. 

Now, it is able to undo all deleted contacts in the same session. I.e. deleting a contact, closing and reopening the app will not be able to undo.
It is also able to undo clear, and multiple edit and add commands.

### CODE REVIEW

File: src/test/java/seedu/address/logic/commands/DeleteCommandTest.java [hunk 1]
```diff
@@ -79,6 +79,11 @@ public void execute_invalidIndexFilteredList_throwsCommandException() {
         assertCommandFailure(deleteCommand, model, Messages.MESSAGE_INVALID_PERSON_DISPLAYED_INDEX);
     }
 
+    @Test
+    public void execute_didNotStoreDeletedPerson_throwsAssertion() {
+
```
Comment (bhnuka) [inline src/test/java/seedu/address/logic/commands/DeleteCommandTest.java:84]: Wait why is this empty haha was it an accident or am I misunderstanding the testcase

File: src/main/java/seedu/address/model/Model.java [hunk 2]
```diff
@@ -103,5 +104,30 @@ public interface Model {
      */
     Person getDeletedPerson();
 
-    void undo();
+    /**
+     * Removes the most recently deleted person from the deleted person arraylist.
+     */
+    void removeDeletedPerson();
+
```
Comment (bhnuka) [inline src/main/java/seedu/address/model/Model.java:111]: Should add javadoc comments for all the methods below

**Review by bhnuka — COMMENTED** [review]
Overall, just one empty testcase I'm confused about and a slight lack of Javadoc comments (should be easy enough to rectify). Other than that, LGTM!

File: src/test/java/seedu/address/logic/commands/DeleteCommandTest.java (hunk 1 above)
Comment (glenngnng) [reply src/test/java/seedu/address/logic/commands/DeleteCommandTest.java:84]: Woops! It was meant to be deleted... Thanks for pointing it out. :D

File: src/main/java/seedu/address/logic/commands/UndoCommand.java [hunk 3]
```diff
@@ -31,17 +37,78 @@ public class UndoCommand extends Command {
...
+            return executeUndoDelete(model);
+        case "clear":
+            return executeUndoClear(model);
+        case "add":
+        case "clone":
+            return executeUndoAdd(model);
+        case "edit":
+            return executeUndoEdit(model);
+        default:
+            return null;
```
Comment (syamfarh) [inline src/main/java/seedu/address/logic/commands/UndoCommand.java]: Does returning null result in any errors to the function it pass the value to? I had problem once with appointment date for returning null values if no dates are indicated

**Review by syamfarh — COMMENTED** [review]
A bit doubtful of null value

File: src/main/java/seedu/address/model/Model.java (hunk 2 above)
Comment (glenngnng) [reply src/main/java/seedu/address/model/Model.java:111]: Added!

File: src/main/java/seedu/address/logic/commands/UndoCommand.java (hunk 3 above)
Comment (glenngnng) [reply src/main/java/seedu/address/logic/commands/UndoCommand.java]: You're right! I'll look into this. I had previously assumed that the default should never occur.

Comment (glenngnng) [reply src/main/java/seedu/address/logic/commands/UndoCommand.java]: I've amended this by throwing an AssertionError for the default case in my latest commit. :D

**Comment by bhnuka** [comment]
Okay LGTM! Merging now

**Pull request merged**